plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMergeSink.java

```diff
@@ -0,0 +1,339 @@
+    {
+        this.hdfsEnvironment = requireNonNull(hdfsEnvironment, "hdfsEnvironment is null");
+        this.session = requireNonNull(session, "session is null");
+        this.parquetDateTimeZone = requireNonNull(parquetDateTimeZone, "parquetDateTimeZone is null");
+        this.trinoVersion = requireNonNull(trinoVersion, "trinoVersion is null");
+        this.dataFileInfoCodec = requireNonNull(dataFileInfoCodec, "dataFileInfoCodec is null");
+        this.mergeResultJsonCodec = requireNonNull(mergeResultJsonCodec, "mergeResultJsonCodec is null");
+        this.writerStats = requireNonNull(writerStats, "writerStats is null");
+        this.rootTableLocation = requireNonNull(rootTableLocation, "rootTableLocation is null");
+        this.insertPageSink = requireNonNull(insertPageSink, "insertPageSink is null");
+        requireNonNull(tableColumns, "tableColumns is null");
+        this.tableColumnCount = tableColumns.size();
+        this.dataColumns = tableColumns.stream()
+                .filter(column -> column.getColumnType() == REGULAR)
+                .collect(toImmutableList());
+    }
+
+    @Override
+    public void storeMergedRows(Page page)
+    {
+        MergePage mergePage = createDeleteAndInsertPages(page, tableColumnCount);
+
+        mergePage.getInsertionsPage().ifPresent(insertPageSink::appendPage);
+
+        mergePage.getDeletionsPage().ifPresent(deletions -> {
+            ColumnarRow rowIdRow = toColumnarRow(deletions.getBlock(deletions.getChannelCount() - 1));
+
+            for (int position = 0; position < rowIdRow.getPositionCount(); position++) {
+                Slice filePath = VARCHAR.getSlice(rowIdRow.getField(0), position);
+                long rowPosition = BIGINT.getLong(rowIdRow.getField(1), position);
+                Slice partitions = VARCHAR.getSlice(rowIdRow.getField(2), position);
+
+                List<String> partitionValues = PARTITIONS_CODEC.fromJson(partitions.toStringUtf8());
+
+                FileDeletion deletion = fileDeletions.computeIfAbsent(filePath, x -> new FileDeletion(partitionValues));
+
+                deletion.rowsToDelete().addLong(rowPosition);
+            }
+        });
+    }
+
+    @Override
+    public CompletableFuture<Collection<Slice>> finish()
+    {
+        List<Slice> fragments = new ArrayList<>();
+
+        insertPageSink.finish().join().stream()
+                .map(Slice::getBytes)
+                .map(dataFileInfoCodec::fromJson)
+                .map(info -> new DeltaLakeMergeResult(Optional.empty(), Optional.of(info)))
+                .map(mergeResultJsonCodec::toJsonBytes)
+                .map(Slices::wrappedBuffer)
+                .forEach(fragments::add);
+
+        fileDeletions.forEach((path, deletion) ->
+                fragments.addAll(rewriteFile(path.toStringUtf8(), deletion)));
+
+        return completedFuture(fragments);
+    }
+
+    private List<Slice> rewriteFile(String filePath, FileDeletion deletion)
```

Comment:
Maybe add a comment about Delta not actually supporting delta deletes currently (despite the name), so we have to rewrite the file

Reply:
Added a comment.